Compilation error caused by data() member

I have an interesting problem. The following code:

```
#include "span.hpp"
#include <vector>

class Test
{
public:
    Test() { }
    Test(const Test&) = default;
    /* explicit */ Test(const tcb::span<const int>&) { } // Making the constructor explicit **doesn't** fix the issue

    Test& operator=(const Test&) = default;
    Test& operator=(const tcb::span<const int>&) { return *this; }

    const std::vector<int>& data() const { return m_data; } // Commenting this member will fix the error

private:
    std::vector<int> m_data;
};

int main()
{
    Test test;
    test = Test();
    return 0;
}
```

Produces the following error on Visual Studio 2019 16.11.7 (any /std:c++xx setting):

```
1>span.hpp(271,1): error C2234: 'abstract declarator': arrays of references are illegal
1>ConsoleApplication1.cpp(28): message : see reference to class template instantiation 'tcb::detail::is_container_element_type_compatible<const Test &,const char,void>' being compiled
```

It also produces a similar error in gcc 9.3.0:
```
/home/ceztko/span.hpp: In instantiation of ‘struct tcb::detail::is_container_element_type_compatible<Test&, const int, void>’:
/home/ceztko/span.hpp:363:75:   required by substitution of ‘template<class Container, long unsigned int E, typename std::enable_if<(((E == tcb::dynamic_extent) && tcb::detail::is_container<Container, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value) && tcb::detail::is_container_element_type_compatible<Container&, const int, void>::value), int>::type <anonymous> > constexpr tcb::span<const int>::span(Container&) [with Container = Test; long unsigned int E = 18446744073709551615; typename std::enable_if<(((E == tcb::dynamic_extent) && tcb::detail::is_container<Container, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value) && tcb::detail::is_container_element_type_compatible<Container&, const int, void>::value), int>::type <anonymous> = <missing>]’
test.c++:23:17:   required from here
/home/ceztko/span.hpp:264:8: error: creating array of ‘std::__remove_pointer_helper<const std::vector<int>&, const std::vector<int>&>::type’ {aka ‘const std::vector<int>&’}
  264 | struct is_container_element_type_compatible<
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  265 |     T, E,
      |     ~~~~~
  266 |     typename std::enable_if<
      |     ~~~~~~~~~~~~~~~~~~~~~~~~
  267 |         !std::is_same<typename std::remove_cv<decltype(
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  268 |                           detail::data(std::declval<T>()))>::type,
      |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  269 |                       void>::value>::type>
      |                       ~~~~~~~~~~~~~~~~~~~~
/home/ceztko/span.hpp: In instantiation of ‘struct tcb::detail::is_container_element_type_compatible<const Test&, const int, void>’:
/home/ceztko/span.hpp:374:75:   required by substitution of ‘template<class Container, long unsigned int E, typename std::enable_if<(((E == tcb::dynamic_extent) && tcb::detail::is_container<Container, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value) && tcb::detail::is_container_element_type_compatible<const Container&, const int, void>::value), int>::type <anonymous> > constexpr tcb::span<const int>::span(const Container&) [with Container = Test; long unsigned int E = 18446744073709551615; typename std::enable_if<(((E == tcb::dynamic_extent) && tcb::detail::is_container<Container, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value) && tcb::detail::is_container_element_type_compatible<const Container&, const int, void>::value), int>::type <anonymous> = <missing>]’
test.c++:23:17:   required from here
/home/ceztko/span.hpp:264:8: error: creating array of ‘std::__remove_pointer_helper<const std::vector<int>&, const std::vector<int>&>::type’ {aka ‘const std::vector<int>&’}
```

If I comment the `data()` member, the problem fixes itself, there's clearly a clash. A similar [code](https://wandbox.org/permlink/1ka9Uta93S1rN0pO) using a supplied `span` in a recent C++20 compliant compiler doesn't show the issue. Fortunately the problem was easy to workaround in my case but a fix would be very welcome since it was quite annoying to debug and it could happen elsewhere, since a `data()` member is to be commonly found.

--- a/include/tcb/span.hpp
+++ b/include/tcb/span.hpp
@@ -264,12 +264,14 @@ template <typename T, typename E>
 struct is_container_element_type_compatible<
     T, E,
     typename std::enable_if<
-        !std::is_same<typename std::remove_cv<decltype(
-                          detail::data(std::declval<T>()))>::type,
-                      void>::value>::type>
-    : std::is_convertible<
-          remove_pointer_t<decltype(detail::data(std::declval<T>()))> (*)[],
-          E (*)[]> {};
+        !std::is_same<
+            typename std::remove_cv<decltype(detail::data(std::declval<T>()))>::type,
+            void>::value &&
+        std::is_convertible<
+            remove_pointer_t<decltype(detail::data(std::declval<T>()))> (*)[],
+            E (*)[]>::value
+        >::type>
+    : std::true_type {};
 
 template <typename, typename = size_t>
 struct is_complete : std::false_type {};

